File: e2e/control-plane/01-Infrastructure/otel-demo/custom/frontend/utils/featbit/FeatBit.ts
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0
//
// FeatBit native server-side SDK integration for the Next.js frontend.
//
// Design goals (see e2e/control-plane continuity/failover testing) — mirrors
// the payment service's featbit.js and the recommendation service's
// featbit_client.py:
//   * Use FeatBit's NATIVE Node SERVER SDK (@featbit/node-server-sdk),
//     evaluated SERVER-SIDE only (in pages/api/*.ts route handlers). This is
//     NOT OpenFeature and NOT the client/browser SDK.
//   * A single FbClient (lazy singleton) for the process lifetime — the
//     FeatBit best practice. The frontend has no explicit boot hook, so the
//     client is initialized on first evaluation and reused thereafter.
//   * Graceful degradation: if FeatBit is unconfigured, the eval/streaming
//     server is unreachable, or a flag is not found, every evaluation returns
//     a SAFE DEFAULT and the storefront keeps rendering. This is what lets us
//     kill the eval-server / break flags and observe continuity. We NEVER
//     throw on a connection failure or a missing flag.
//   * BUT a *misconfigured* flag value (e.g. an unknown checkout button
//     variant) is returned faithfully to the caller. Surfacing how the
//     application copes with bad-but-present flag values is the whole point of
//     the resiliency exercise, so we do NOT sanitize those here. See
//     pages/api/checkout.ts for the intentional unknown-variant gap.
//
// TypeScript / ESM module. Server-only — never import this from a component
// that runs in the browser.

import { FbClientBuilder, UserBuilder } from '@featbit/node-server-sdk';

// Minimal structural view of the SDK surface we use. Declaring it locally keeps
// the Next.js build robust regardless of the exact type names the SDK exports,
// while still giving us type-checked evaluation calls.
interface FbClientLike {
  waitForInitialization(): Promise<void>;
  boolVariation(key: string, user: unknown, defaultValue: boolean): Promise<boolean>;
  numberVariation(key: string, user: unknown, defaultValue: number): Promise<number>;
  stringVariation(key: string, user: unknown, defaultValue: string): Promise<string>;
  close(): Promise<void>;
}

// FeatBit evaluation context. The `key` must stably and uniquely identify the
// evaluation subject; for service-scoped operational/release/experiment flags
// the subject is the frontend service itself.
const SERVICE_USER: unknown = new UserBuilder('frontend').name('frontend').build();

// Flag keys (kebab-case, intention-revealing, one concern each).
export const FLAG_RELEASE_REDESIGNED_PRODUCT_PAGE = 'release-redesigned-product-page'; // release gate (bool)
export const FLAG_RECOMMENDATIONS_ENABLED = 'operational-recommendations-enabled';     // ops toggle   (bool)
export const FLAG_CHECKOUT_BUTTON_VARIANT = 'experiment-checkout-button-variant';      // experiment   (string)

// Safe defaults — preserve the storefront's normal behavior when FeatBit can't
// be reached or a flag is absent.
export const DEFAULT_RELEASE_REDESIGNED_PRODUCT_PAGE = false;
export const DEFAULT_RECOMMENDATIONS_ENABLED = true;
export const DEFAULT_CHECKOUT_BUTTON_VARIANT = 'control';

// Lazy singleton state.
let client: FbClientLike | null = null;
let initStarted = false;
let initPromise: Promise<void> = Promise.resolve();

/**
 * Initialize a single FeatBit client for the process lifetime.
 *
 * Never throws. When FeatBit is unconfigured we leave `client` null and every
 * evaluation falls back to its safe default. When configured we keep the built
 * client reference even if waitForInitialization rejects — the SDK keeps
 * reconnecting in the background, and until it is connected evaluations return
 * their defaults (graceful failover) rather than crashing the storefront.
 */
async function init(): Promise<void> {
  const envSecret = process.env.FEATBIT_ENV_SECRET;
  // Accept either FEATBIT_EVENT_URL or FEATBIT_EVAL_URL for the events endpoint.
  const eventUrl = process.env.FEATBIT_EVENT_URL || process.env.FEATBIT_EVAL_URL;
  const streamingUrl = process.env.FEATBIT_STREAMING_URL;

  if (!(envSecret && eventUrl && streamingUrl)) {
    console.info(
      '[featbit] not configured (need FEATBIT_ENV_SECRET, ' +
        'FEATBIT_EVENT_URL/FEATBIT_EVAL_URL, FEATBIT_STREAMING_URL); ' +
        'frontend flags will use safe defaults'
    );
    return;
  }

  try {
    const built = new FbClientBuilder()
      .sdkKey(envSecret)
      .streamingUri(streamingUrl)
      .eventsUri(eventUrl)
      .build() as unknown as FbClientLike;

    // Keep the reference immediately so background reconnection benefits later
    // evaluations even if this initial wait times out.
    client = built;

    // waitForInitialization may reject if the eval-server is unreachable. We
    // catch it below: evaluations return defaults until the SDK connects.
    await built.waitForInitialization();
    console.info(`[featbit] client initialized (event=${eventUrl} streaming=${streamingUrl})`);
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    console.warn(`[featbit] init failed (${message}); frontend flags will use safe defaults`);
  }
}

/** Kick off (once) and return the shared initialization promise. */
function ensureInitialized(): Promise<void> {
  if (!initStarted) {
    initStarted = true;
    initPromise = init();
  }
  return initPromise;
}

/**
 * Evaluate a flag, always returning a usable value.
 *
 * flag-not-found, client-not-ready, and eval-server-down all return `def`
 * (non-breaking). A present-but-misconfigured value is returned as-is.
 */
async function variation<T>(
  evalFn: (c: FbClientLike, user: unknown, def: T) => Promise<T>,
  key: string,
  def: T
): Promise<T> {
  await ensureInitialized();
  if (client === null) {
    return def;
  }
  try {
    return await evalFn(client, SERVICE_USER, def);
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    console.warn(`[featbit] evaluation error for '${key}' (${message}); using default`);
    return def;
  }
}

/**
 * bool — release gate for the redesigned product page layout.
 * Safe either way; default OFF keeps the current layout.
 */
export async function getReleaseRedesignedProductPage(): Promise<boolean> {
  return variation<boolean>(
    (c, u, d) => c.boolVariation(FLAG_RELEASE_REDESIGNED_PRODUCT_PAGE, u, d),
    FLAG_RELEASE_REDESIGNED_PRODUCT_PAGE,
    DEFAULT_RELEASE_REDESIGNED_PRODUCT_PAGE
  );
}

/**
 * bool — operational kill switch for recommendations. When false the
 * recommendations route returns an empty list (graceful degradation, e.g. when
 * the recommendation service is unhealthy). Default ON.
 */
export async function getRecommendationsEnabled(): Promise<boolean> {
  return variation<boolean>(
    (c, u, d) => c.boolVariation(FLAG_RECOMMENDATIONS_ENABLED, u, d),
    FLAG_RECOMMENDATIONS_ENABLED,
    DEFAULT_RECOMMENDATIONS_ENABLED
  );
}

/**
 * string — A/B experiment selecting the checkout button variant
 * (control | variant-a | variant-b). Returned VERBATIM and intentionally NOT
 * validated here. An unrecognized variant is handled by an unchecked lookup in
 * checkout.ts and surfaces a real error (intentional unknown-variant gap).
 */
export async function getCheckoutButtonVariant(): Promise<string> {
  return variation<string>(
    (c, u, d) => c.stringVariation(FLAG_CHECKOUT_BUTTON_VARIANT, u, d),
    FLAG_CHECKOUT_BUTTON_VARIANT,
    DEFAULT_CHECKOUT_BUTTON_VARIANT
  );
}

/** Best-effort shutdown (exported for completeness / tests). */
export async function close(): Promise<void> {
  if (client !== null) {
    try {
      await client.close();
    } catch {
      // best-effort shutdown
    }
  }
}

File: e2e/control-plane/01-Infrastructure/otel-demo/custom/checkout/featbit.go
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0
//
// FeatBit native server-side SDK integration for the checkout service.
//
// Design goals (see e2e/control-plane continuity/failover testing):
//   - Use FeatBit's NATIVE Go SDK (github.com/featbit/featbit-go-sdk), NOT
//     OpenFeature. (The upstream service keeps its own OpenFeature/flagd wiring
//     for the unrelated demo chaos flags like paymentUnreachable and
//     kafkaQueueProblems; this file is an independent, additive FeatBit client.)
//   - One client for the process lifetime (FeatBit best practice).
//   - Graceful degradation: if FeatBit is unconfigured, the eval/streaming
//     server is unreachable, or a flag is not found, every evaluation returns a
//     SAFE DEFAULT and the service keeps working. This is what lets us kill the
//     eval-server / break flags and observe checkout continuity.
//   - BUT a *misconfigured* flag value (an unrecognized shipping strategy) is
//     returned faithfully to the caller. Surfacing how the application copes
//     with bad-but-present flag values is the whole point of the resiliency
//     exercise, so we do NOT sanitize those here — the caller's switch turns an
//     unknown value into a real gRPC error.
package main

import (
	"fmt"
	"os"

	featbit "github.com/featbit/featbit-go-sdk"
	"github.com/featbit/featbit-go-sdk/interfaces"
)

// Flag keys (kebab-case, intention-revealing, one concern each).
const (
	flagCheckoutEnabled  = "operational-checkout-enabled"      // operational (bool)
	flagExpressCheckout  = "release-express-checkout"          // release     (bool)
	flagShippingStrategy = "experiment-shipping-cost-strategy" // experiment  (string)
)

// Safe defaults — preserve the service's normal behavior when FeatBit can't be
// reached or a flag is absent.
const (
	defaultCheckoutEnabled  = true       // checkout open for business
	defaultExpressCheckout  = false      // express/one-click path off
	defaultShippingStrategy = "standard" // ordinary shipping quote, unchanged
)

// featBit wraps a single FeatBit client plus the service-scoped evaluation user.
// A nil client means "disabled mode": every getter returns its safe default and
// the service runs with NO FeatBit backend.
type featBit struct {
	client *featbit.FBClient
	user   interfaces.FBUser
}

// fb is the process-wide FeatBit handle. It is never nil after initFeatBit():
// in disabled mode fb.client is nil and all getters fall back to safe defaults.
var fb = &featBit{}

// initFeatBit initializes a single FeatBit client for the process lifetime.
//
// It reads FEATBIT_ENV_SECRET, FEATBIT_EVENT_URL (or FEATBIT_EVAL_URL) and
// FEATBIT_STREAMING_URL. If any are unset, or the client cannot be created at
// all (invalid secret/URL), the service operates in disabled mode (safe
// defaults). If the eval-server is merely unreachable at startup, the SDK still
// returns a usable client that serves defaults and reconnects in the
// background — so flags "heal" automatically once connectivity returns.
func initFeatBit() {
	envSecret := os.Getenv("FEATBIT_ENV_SECRET")
	// Accept either FEATBIT_EVENT_URL or FEATBIT_EVAL_URL for the events endpoint.
	eventURL := os.Getenv("FEATBIT_EVENT_URL")
	if eventURL == "" {
		eventURL = os.Getenv("FEATBIT_EVAL_URL")
	}
	streamingURL := os.Getenv("FEATBIT_STREAMING_URL")

	if envSecret == "" || eventURL == "" || streamingURL == "" {
		logger.Info("FeatBit not configured (need FEATBIT_ENV_SECRET, " +
			"FEATBIT_EVENT_URL/FEATBIT_EVAL_URL, FEATBIT_STREAMING_URL); " +
			"checkout flags will use safe defaults")
		return
	}

	// Service-scoped evaluation subject. The key must stably and uniquely
	// identify the subject; for service-scoped operational/release flags the
	// subject is the checkout service itself.
	user, err := interfaces.NewUserBuilder("checkout").
		UserName("checkout").
		Build()
	if err != nil {
		logger.Warn(fmt.Sprintf("FeatBit user build failed (%v); "+
			"checkout flags will use safe defaults", err))
		return
	}

	// NewFBClient(envSecret, streamingUrl, eventUrl). Note the streaming/event
	// argument order. On a connect timeout or "initialization failed" the SDK
	// still returns a NON-NIL client that serves defaults and keeps retrying;
	// it returns nil ONLY when the config is invalid (bad secret/URL).
	client, err := featbit.NewFBClient(envSecret, streamingURL, eventURL)
	if err != nil {
		if client == nil {
			logger.Warn(fmt.Sprintf("FeatBit client could not be created (%v); "+
				"checkout flags will use safe defaults", err))
			return
		}
		logger.Warn(fmt.Sprintf("FeatBit client not yet connected (%v); "+
			"evaluations use defaults until the eval-server is reachable", err))
	}

	fb.client = client
	fb.user = user
	if client.IsInitialized() {
		logger.Info(fmt.Sprintf("FeatBit client initialized (streaming=%s event=%s)",
			streamingURL, eventURL))
	}
}

// close shuts down the FeatBit client (flushes pending events). Safe to call in
// disabled mode.
func (f *featBit) close() {
	if f != nil && f.client != nil {
		_ = f.client.Close()
	}
}

// boolVariation evaluates a boolean flag, never panicking. flag-not-found,
// client-not-ready and eval-server-down all return def (non-breaking).
func (f *featBit) boolVariation(key string, def bool) (val bool) {
	val = def
	if f == nil || f.client == nil {
		return
	}
	defer func() {
		if r := recover(); r != nil {
			logger.Warn(fmt.Sprintf("FeatBit eval panic for %q (%v); using default", key, r))
			val = def
		}
	}()
	v, _, err := f.client.BoolVariation(key, f.user, def)
	if err != nil {
		logger.Warn(fmt.Sprintf("FeatBit eval error for %q (%v); using default", key, err))
		return def
	}
	return v
}

// stringVariation evaluates a string flag, never panicking. A present-but-bad
// value (e.g. an unrecognized shipping strategy) is returned VERBATIM — the
// caller is responsible for how it copes, which is the point of the resiliency
// exercise.
func (f *featBit) stringVariation(key string, def string) (val string) {
	val = def
	if f == nil || f.client == nil {
		return
	}
	defer func() {
		if r := recover(); r != nil {
			logger.Warn(fmt.Sprintf("FeatBit eval panic for %q (%v); using default", key, r))
			val = def
		}
	}()
	v, _, err := f.client.Variation(key, f.user, def)
	if err != nil {
		logger.Warn(fmt.Sprintf("FeatBit eval error for %q (%v); using default", key, err))
		return def
	}
	return v
}

// checkoutEnabled is the master operational kill-switch. When it evaluates to
// false the caller returns a clean gRPC error instead of placing an order; this
// is deliberate operational behavior, NOT a crash.
func (f *featBit) checkoutEnabled() bool {
	return f.boolVariation(flagCheckoutEnabled, defaultCheckoutEnabled)
}

// expressCheckoutEnabled gates the (simple, safe) express/one-click code path.
func (f *featBit) expressCheckoutEnabled() bool {
	return f.boolVariation(flagExpressCheckout, defaultExpressCheckout)
}

// shippingCostStrategy selects the A/B shipping cost strategy. Returned VERBATIM
// with NO validation here, so an unrecognized value reaches the caller's switch
// and produces a real error for the gRPC caller (intentional resiliency gap).
func (f *featBit) shippingCostStrategy() string {
	return f.stringVariation(flagShippingStrategy, defaultShippingStrategy)
}
